neuraxle/metaopt/callbacks.py

```diff
@@ -0,0 +1,250 @@
+        pass
+
+
+class BaseRefitCallback(ABC):
+    @abstractmethod
+    def call(self, scores):
+        pass
+
+
+class EarlyStoppingCallback(BaseCallback):
+    """
+    Perform early stopping when there is multiple epochs in a row that didn't improve the performance of the model.
+
+    Example usage :
+
+    .. code-block:: python
+
+        trainer = Trainer(
+            metrics=self.metrics,
+            callbacks=self.callbacks,
+            score=self.scoring_function,
+            epochs=self.epochs
+        )
+
+        trial = trainer.execute_trial(
+            p=p,
+            trial_repository=repo_trial,
+            train_data_container=training_data_container,
+            validation_data_container=validation_data_container,
+            context=context
+        )
+
+        pipeline = trainer.refit(repo_trial.pipeline, data_container, context)
+
+
+    .. seealso::
+        :class:`AutoML`,
+        :class:`Trial`,
+        :class:`HyperparamsRepository`,
+        :class:`HyperparameterOptimizer`,
+        :class:`RandomSearchHyperparameterOptimizer`,
+        :class:`DataContainer`
+    """
+
+    def __init__(self, n_epochs_without_improvement, higher_score_is_better):
+        self.higher_score_is_better = higher_score_is_better
+        self.n_epochs_without_improvement = n_epochs_without_improvement
+
+    def call(
+            self,
+            trial: Trial,
+            epoch_number: int,
+            total_epochs: int,
+            input_train: DataContainer,
+            pred_train: DataContainer,
+            input_val: DataContainer,
+            pred_val: DataContainer,
+            is_finished_and_fitted: bool
+    ):
+        validation_scores = trial.get_validation_scores()
+        if len(validation_scores) > self.n_epochs_without_improvement:
+            if validation_scores[-self.n_epochs_without_improvement] >= validation_scores[-1] and self.higher_score_is_better:
+                return True
+            if validation_scores[-self.n_epochs_without_improvement] <= validation_scores[-1] and not self.higher_score_is_better:
+                return True
+        return False
+
+
+class MetaCallback(BaseCallback):
+    def __init__(self, wrapped_callback: BaseCallback):
+        self.wrapped_callback = wrapped_callback
+
+    @abstractmethod
+    def call(
+            self,
+            trial: Trial,
+            epoch_number: int,
+            total_epochs: int,
+            input_train: DataContainer,
+            pred_train: DataContainer,
+            input_val: DataContainer,
+            pred_val: DataContainer,
+            is_finished_and_fitted: bool
+    ):
+        pass
+
+class IfBestScore(MetaCallback):
+    def call(self, trial: Trial, epoch_number: int, total_epochs: int, input_train: DataContainer,
+             pred_train: DataContainer, input_val: DataContainer, pred_val: DataContainer, is_finished_and_fitted: bool):
+        if trial.is_new_best_score():
```

Comment:
I wouldn't want to force to sort the callbacks in a certain order. This if seems like it could be missed an only catch in the next iter if not put after the score callback.

Reply:
https://github.com/Neuraxio/Neuraxle/issues/287